From d47d48c3045892e57ad2695ee429683756db0192 Mon Sep 17 00:00:00 2001
From: Koen Kooi <koen@dominion.thruhere.net>
Date: Wed, 8 Jul 2026 16:54:34 +0200
Subject: [PATCH 26/46] AnnaThermostat: prefer direct corrected temperature

Use the temperature point_log as authoritative for room temperature, and only fall back to uncorrected_temperature + temperature_offset when temperature is missing.

Signed-off-by: Koen Kooi <koen@dominion.thruhere.net>
Assisted-by: GPT-5.3-Codex

Upstream-Status: Inappropriate [WIP]
---
 hardware/AnnaThermostat.cpp | 54 +++++++++++++++++++++++++++++++++++--
 1 file changed, 52 insertions(+), 2 deletions(-)

diff --git a/hardware/AnnaThermostat.cpp b/hardware/AnnaThermostat.cpp
index 796c2204b..9589dbe99 100644
--- a/hardware/AnnaThermostat.cpp
+++ b/hardware/AnnaThermostat.cpp
@@ -926,12 +926,25 @@ void CAnnaThermostat::GetMeterDetails()
 		}
 
 		std::vector<std::pair<std::string, std::string>> pointLogs;
+		float reportedTemperature = 0.0F;
+		float uncorrectedTemperature = 0.0F;
+		float temperatureOffset = 0.0F;
+		bool hasReportedTemperature = false;
+		bool hasUncorrectedTemperature = false;
+		bool hasTemperatureOffset = false;
 		for (TiXmlElement* pLogElem = pElem; pLogElem; pLogElem = pLogElem->NextSiblingElement())
 		{
 			sname = GetElementChildValue(pLogElem, "type");
 			tmpstr = GetPeriodMeasurement(pLogElem);
 			pointLogs.emplace_back(sname, tmpstr);
 
+			if (sname == "temperature" && !tmpstr.empty())
+				hasReportedTemperature = isNumericValue(tmpstr, reportedTemperature);
+			else if (sname == "uncorrected_temperature" && !tmpstr.empty())
+				hasUncorrectedTemperature = isNumericValue(tmpstr, uncorrectedTemperature);
+			else if (sname == "temperature_offset" && !tmpstr.empty())
+				hasTemperatureOffset = isNumericValue(tmpstr, temperatureOffset);
+
 			if (sname == "battery" && !tmpstr.empty())
 			{
 				float batteryRaw = (float)atof(tmpstr.c_str());
@@ -945,6 +958,21 @@ void CAnnaThermostat::GetMeterDetails()
 			}
 		}
 
+		bool hasEffectiveTemperature = false;
+		float effectiveTemperature = 0.0F;
+		if (hasReportedTemperature)
+		{
+			effectiveTemperature = reportedTemperature;
+			hasEffectiveTemperature = true;
+		}
+		else if (hasUncorrectedTemperature && hasTemperatureOffset)
+		{
+			effectiveTemperature = uncorrectedTemperature + temperatureOffset;
+			hasEffectiveTemperature = true;
+		}
+
+		bool temperaturePublished = false;
+
 		for (const auto& pointLog : pointLogs)
 		{
 			sname = pointLog.first;
@@ -961,10 +989,11 @@ void CAnnaThermostat::GetMeterDetails()
 
 			if (sname == "temperature")
 			{
-				if (!tmpstr.empty())
+				if (!tmpstr.empty() || hasEffectiveTemperature)
 				{
-					float temperature = (float)atof(tmpstr.c_str());
+					float temperature = hasEffectiveTemperature ? effectiveTemperature : (float)atof(tmpstr.c_str());
 					SendTempSensor(applianceIDhex, batterypercentage, temperature, ApplianceName + " " + sname);;
+					temperaturePublished = true;
 				}
 			}
 			else if (sname == "illuminance")
@@ -1023,6 +1052,22 @@ void CAnnaThermostat::GetMeterDetails()
 					SendTempSensor(applianceIDhex + 4, 255, temperature, ApplianceName + " " + sname);;
 				}
 			}
+			else if (sname == "temperature_offset")
+			{
+				if (!tmpstr.empty())
+				{
+					float temperature = (float)atof(tmpstr.c_str());
+					SendTempSensor(applianceIDhex + 7, batterypercentage, temperature, ApplianceName + " " + sname);
+				}
+			}
+			else if (sname == "temperature_difference")
+			{
+				if (!tmpstr.empty())
+				{
+					float temperature = (float)atof(tmpstr.c_str());
+					SendTempSensor(applianceIDhex + 8, batterypercentage, temperature, ApplianceName + " " + sname);
+				}
+			}
 			else if (sname == "boiler_state")
 			{
 				if (!tmpstr.empty())
@@ -1304,6 +1349,11 @@ void CAnnaThermostat::GetMeterDetails()
 				}
 			}
         }
+
+		if (!temperaturePublished && hasEffectiveTemperature)
+		{
+			SendTempSensor(applianceIDhex, batterypercentage, effectiveTemperature, ApplianceName + " temperature");
+		}
         pAppliance = pAppliance->NextSiblingElement("appliance");
     }
 
